File: CS50/week_1/credit/credit.c
#include <stdio.h>
#include <stdlib.h>

int getSize(const char *str)
{
    int size = 0;
    for (int i = 0; str[i] != '\0'; i++)
        size++;
    return size;
}

int isNumeric(const char *str)
{
    for (int i = 0; i < getSize(str); i++)
    {
        if (str[i] < '0' || str[i] > '9')
        {
            return 0;
        }
    }
    return 1;
}

void luhnsCheck(const char *num)
{
    int sum = 0;
    int size = getSize(num);
    for (int i = size - 1; i >= 0; i -= 2)
    {
        sum += num[i] - '0';
        int val = (num[i - 1] - '0') * 2;
        if (val == 0 || i == 0)
            continue;
        char *str = malloc(5 * sizeof(char));
        sprintf(str, "%d", val);
        int valSize = getSize(str);
        for (int j = 0; j < valSize; j++)
        {
            sum += (str[j] - '0');
        }
    }

    if (sum % 10)
    {
        printf("INVALID\n");
    }
    else if (size == 15 && num[0] == '3' && (num[1] == '4' || num[1] == '7'))
    {
        printf("AMEX\n");
    }
    else if ((size == 13 || size == 16) && num[0] == '4')
    {
        printf("VISA\n");
    }
    else if (size == 16 && num[0] == '5' && num[1] > '0' && num[1] < '6')
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

int main(void)
{
    char *num = malloc(50 * sizeof(char));
    do
    {
        printf("Enter your card number:\n>> ");
        scanf("%s", num);
    }
    while (!isNumeric(num));
    luhnsCheck(num);
    free(num);
}

File: CS50/week_6/sentimental-credit/credit.py
def validate(number: str) -> bool:
    total: int = 0
    mode: int = 0
    for char in number[::-1]:
        if mode == 0:
            total += int(char)
            mode = 1
        else:
            total += sum([int(digit) for digit in str(int(char) * 2)])
            mode = 0
    return not total % 10


def check_card(number: str) -> str:
    size: int = len(number)

    if not validate(number):
        return "INVALID"

    elif size == 15 and number[:2] in ["34", "37"]:
        return "AMEX"

    elif size == 16 and int(number[:2]) in range(50, 56):
        return "MASTERCARD"

    elif size in [13, 16] and number[0] == "4":
        return "VISA"

    return "INVALID"


def main() -> None:
    while True:
        number: str = input("Number: ")
        if number.isdigit():
            break
    print(check_card(number))


if __name__ == "__main__":
    main()
